File: cs/Parking/Lot/ParkingLot.cs
using System;
using System.Collections.Generic;

namespace Parking
{
    public class ParkingLot
    {
        public ParkingLotID ID { get; private set; }
        public int AvailableSpaces { get; private set; }
        public bool HasAvailableSpaces => AvailableSpaces > 0;

        private readonly IDictionary<Ticket, Car> parkedCars;

        public ParkingLot(int totalPosition)
        {
            ID = new ParkingLotID(Guid.NewGuid().ToString());
            AvailableSpaces = totalPosition;
            parkedCars = new Dictionary<Ticket, Car>();
        }

        public Ticket Park(Car car)
        {
            if (AvailableSpaces <= 0)
            {
                throw new Exception("capacity is not enough.");
            }

            var ticket = new Ticket(car.PlateNumber, ID);
            parkedCars.Add(ticket, car);
            AvailableSpaces--;
            return ticket;
        }

        public Car Take(Ticket ticket)
        {
            if (!ticket.IsValid || !parkedCars.ContainsKey(ticket))
            {
                throw new Exception("Ticket is invalided.");
            }

            var car = parkedCars[ticket];
            ticket.Invalidate();
            parkedCars.Remove(ticket);
            AvailableSpaces++;
            return car;
        }

    }

    public class ParkingLotID : ValueObject<string>
    {
        public ParkingLotID(string parkingLotID) : base(parkingLotID) { }
    }
}

File: kt/src/main/kotlin/com/github/scglwsj/parking/lot/ParkingLot.kt
package com.github.scglwsj.parking.lot

import java.lang.Exception
import java.util.*

class ParkingLot(totalPosition: Int) {
    val id: ParkingLotID =
        ParkingLotID(UUID.randomUUID().toString())
    var availableSpaces = totalPosition
        private set
    private val parkedCars = mutableMapOf<Ticket, Car>()

    fun park(car: Car): Ticket {
        if (availableSpaces <= 0)
            throw Exception("capacity is not enough.")

        return Ticket(car.plant, id).also {
            parkedCars[it] = car
            availableSpaces--
        }
    }

    fun take(ticket: Ticket): Car {
        if (!ticket.isValid || !parkedCars.containsKey(ticket))
            throw Exception("Ticket is invalided.")

        return parkedCars[ticket]!!.also {
            ticket.invalidate()
            parkedCars.remove(ticket)
            availableSpaces++
        }
    }

    val hasAvailableSpaces: Boolean get() = availableSpaces > 0
}

data class ParkingLotID(val id: String)
